### PR TITLE
Feature/st 108 customization parameters

### PR DESCRIPTION
**Pull request by james-simpson-ST**

### CODE REVIEW

File: src/core/shared/Frame.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { Styler, AllowedStyles, Styles } from './Styler';
+
+export default class Frame {
+
+    constructor() {
+    }
```
Comment (pgs-gmartowski) [inline src/core/shared/Frame.ts]: empty constructor

File: src/core/shared/Frame.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    public onInit() {
+        this._applyStyles();
+    }
+    
+    public parseUrl() {
+        var parsedUrl = new URL(window.location.href);
```
Comment (pgs-gmartowski) [inline src/core/shared/Frame.ts]: we should avoid using `var`, in this case you can use `const`

File: src/core/shared/Frame.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        const styles: Styles = {};
+        parsedUrl.searchParams.forEach(function (value, param) {
```
Comment (pgs-gmartowski) [inline src/core/shared/Frame.ts]: an arrow function can be used here - `parsedUrl.searchParams.forEach((value, param) => {`

File: src/core/shared/Styler.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            templates.push(`${tag} { ${tagStyle} }`);
+        }
+        return templates.join(" ");
+    }
+
+    public inject(styles: Styles) {
+        let style = document.createElement('style');
+        style.innerHTML = this._getStyleString(styles);
+        document.head.appendChild(style);
+    }
```
Comment (pgs-gmartowski) [inline src/core/shared/Styler.ts]: I think this method should be in `DomMethods.class.ts`

Comment (james-simpson-ST) [reply src/core/shared/Styler.ts]: Done

File: src/core/shared/Frame.ts (hunk 3 above)
Comment (james-simpson-ST) [reply src/core/shared/Frame.ts]: Done

File: src/core/shared/Frame.ts (hunk 2 above)
Comment (james-simpson-ST) [reply src/core/shared/Frame.ts]: Done

File: src/core/shared/Frame.ts (hunk 1 above)
Comment (james-simpson-ST) [reply src/core/shared/Frame.ts]: Removed

File: example/index.ts [hunk 5]
```diff
@@ -16,7 +16,12 @@ import { ST } from '../src/stjs';
       controlFrame: 'st-control-frame'
     },
     'st-notification-frame',
-    {},
+    {'background-color-input': 'AliceBlue',
+     'color-input-error': '#721c24',
+     'line-height-input': '12px',
+     'font-size-input': '12px',
```
Comment (pgs-kzasada) [inline example/index.ts]: With this solution, it is impossible to define different styles, e.g. font-size, for different fields. I think we should allow the merchant to style each element as he wants. Maybe for now we can leave it as it is, but we should have in mind that this is not so flexible.

File: src/components/control-frame/ControlFrame.ts [hunk 6]
```diff
@@ -33,11 +35,37 @@ export default class ControlFrame {
...
-  private onInit() {
+  public onInit() {
+    super.onInit();
     this.initEventListeners();
     this.initSubscriptions();
   }
 
+  protected _getAllowedStyles () {
+    let allowed = super._getAllowedStyles();
+    let button = '#' + Selectors.CONTROL_FRAME_BUTTON_SELECTOR;
```
Comment (pgs-kzasada) [inline src/components/control-frame/ControlFrame.ts]: You can use ES2015 Template Strings

File: src/core/Element.ts [hunk 7]
```diff
@@ -64,9 +66,14 @@ export default class Element {
    * @param elementName Name of input which we want to create
    */
 
-  public create(elementName: string) {
+  public create(elementName: string, styles: Styles) {
     this._name = elementName;
-    this._iframeSrc = Element.getComponentAddress(elementName);
+    let address = Element.getComponentAddress(elementName);
+    let params = new URLSearchParams();
```
Comment (pgs-kzasada) [inline src/core/Element.ts]: URLSearchParams are not supported by Internet Explorer, so we have to think about some polyfill.

Comment (aaron-moore-ST) [reply src/core/Element.ts]: Any idea what the compatability is like for the URL interface (eg. `const parsedUrl = new URL('http://www.example.com/demonstration.html');`)? I believe it's older and should be more widely supported.

File: example/index.ts (hunk 5 above)
Comment (james-simpson-ST) [reply example/index.ts]: Added to Jira backlog on task ST-158

File: src/core/Element.ts (hunk 7 above)
Comment (pgs-kzasada) [reply src/core/Element.ts]: `Window.URL` is also not supported by IE. I'll think about some workaround.

Comment (james-simpson-ST) [reply src/core/Element.ts]: @pgs-kzasada I have been looking into this and have been hitting a bigger problem, it appears as if our current webpack/babel config does not work with IE

The problem seems to be the package I've used for the currency formatting on a previous task (money-ts) where this is not being transpiled by babel into ES2015, as a result the code currently crashes out in IE before I can even get to the URL code to try out a polyfill for that. This is already an issue on the develop branch :( 

Do you think you will be able to investigate this or do you think we should have a task at the top of next sprint to sort out compatibility within IE?

File: src/components/control-frame/ControlFrame.ts [hunk 8]
```diff
@@ -33,11 +35,37 @@ export default class ControlFrame {
...
+    let buttonHover = button + ':hover';
+    allowed = { ...allowed,
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/control-frame/ControlFrame.ts]: I feel it is better and more readable to see what we have actually set if we are explicit here, adding in more logic here feels a bit unnecessary when it wants to behave as a whitelist of only allow the merchant to override these specific styles, perhaps if this grew any more we'd want to do so but for now I feel this is manageable

File: src/components/notification-frame/NotificationFrame.ts [hunk 9]
```diff
@@ -48,13 +51,64 @@ export default class NotificationFrame {
...
   }
 
-  private static ELEMENT_ID: string = 'st-notification-frame';
+  protected _getAllowedStyles () {
+    let allowed = super._getAllowedStyles();
+    let notification = '#' + NotificationFrame.ELEMENT_ID;
+    let error = '.' + NotificationFrame.ELEMENT_CLASSES.error + notification;
+    let success = '.' + NotificationFrame.ELEMENT_CLASSES.success + notification;
+    let warning = '.' + NotificationFrame.ELEMENT_CLASSES.warning + notification;
+    let info = '.' + NotificationFrame.ELEMENT_CLASSES.info + notification;
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/components/notification-frame/NotificationFrame.ts]: Done

File: src/components/control-frame/ControlFrame.ts (hunk 6 above)
Comment (james-simpson-ST) [reply src/components/control-frame/ControlFrame.ts]: Done

File: src/core/Element.ts (hunk 7 above)
Comment (pgs-kzasada) [reply src/core/Element.ts]: @james-simpson-ST I've created a separate task for this: STJS-168.

**Pull request merged**